Add a section to explain declarative vs imperative style (Fixes #323) (#429)
cucumber/docs: content/docs/bdd/better-gherkin.md

If the answer is “Yes”, then you should rework it avoiding implementation specific details.
As a side benefit, in consequence your scenarios will be a lot shorter and much easier to follow and understand.

# Consider a more declarative style

> One way to make tests easier to maintain and less brittle is to use a declarative style. Declarative style describes the behaviour of the application, rather than the implementation details. Declarative tests read better as "living documentation". A declarative style helps you focus on the value that the customer is getting, rather than the keystrokes they will use.

Reviewer: ```suggestion
One way to make scenarios easier to maintain and less brittle is to use a declarative style. Declarative style describes the behaviour of the application, rather than the implementation details. Declarative tests read better as "living documentation". A declarative style helps you focus on the value that the customer is getting, rather than the keystrokes they will use.
```

Author: Done.

> One way to make scenarios easier to maintain and less brittle is to use a declarative style. Declarative style describes the behaviour of the application, rather than the implementation details. Declarative tests read better as "living documentation". A declarative style helps you focus on the value that the customer is getting, rather than the keystrokes they will use.

Reviewer: ```suggestion
One way to make scenarios easier to maintain and less brittle is to use a declarative style. Declarative style describes the behaviour of the application, rather than the implementation details. Declarative scenarios read better as "living documentation". A declarative style helps you focus on the value that the customer is getting, rather than the keystrokes they will use.
```

> Imperative tests are communicative, and in some contexts, this style of test is appropriate. On the other hand, because they are so closely tied to the mechanics of the current UI, they often require more work to maintain. Any time the implementation changes, the tests need to be updated too.

Reviewer: ```suggestion
Imperative tests communicate details, and in some contexts, this style of test is appropriate. On the other hand, because they are so closely tied to the mechanics of the current UI, they often require more work to maintain. Any time the implementation changes, the tests need to be updated too.
```

Author: Done. Why does the 'and' show up in red there? Is that something I need to change also?

Reviewer: I don't know. You can ignore it

> Imperative tests communicate details, and in some contexts, this style of test is appropriate. On the other hand, because they are so closely tied to the mechanics of the current UI, they often require more work to maintain. Any time the implementation changes, the tests need to be updated too.

Reviewer: ```suggestion
Imperative tests communicate details, and in some contexts this style of test is appropriate. On the other hand, because they are so closely tied to the mechanics of the current UI, they often require more work to maintain. Any time the implementation changes, the tests need to be updated too.
```

> Here's a more imperative style:

Reviewer: ```suggestion
Here's an example of a feature in an imperative style:
```

Author: Done, and next one I will try to do the "commit changes" thing in the dropdown. Doh.

Reviewer: Learn something new every day ;)

  Given users with a free subscription can access "FreeArticle1" but not "PaidArticle1" 
  When I type "freeFrieda@example.com" in the email field
  And I type "validPassword123" in the password field
  And I press the "Submit" button
  Then I see "FreeArticle1" on the home page
  And I do not see "PaidArticle1" on the home page

Scenario: Subscriber with a paid subscription can access "FreeArticle1" and "PaidArticle1"
  Given I am on the login page
  When I type "paidPattya@example.com" in the email field
  And I type "validPassword123" in the password field
  And I press the "Submit" button
  Then I see "FreeArticle1" and "PaidArticle1" on the home page  
```

Each step is a precise instruction. The inputs and expected results are specified exactly. But it's easy to imagine changes to the application which would require changing these tests. The available options for free versus paid subscriptions can change. Even the means of logging in could change. What if, in the future, users log in with a voice interface or a thumbprint? 

A more declarative style hides the details of how the application's capabilities are implemented.

```

>   Given Free Frieda has a free subscription

Reviewer: ```suggestion
Feature: Subscribers see different sets of stock images based on their subscription level
 
Scenario: Free subscribers see only the free articles
  Given Free Frieda has a free subscription
```

  When Free Frieda logs in with her valid credentials
  Then she sees a Free article on the home page

>   Given Paid Patty has a basic-level paid subscription

Reviewer: ```suggestion
Scenario: Subscriber with a paid subscription can access both free and paid articles
  Given Paid Patty has a basic-level paid subscription
```

  When Paid Patty logs in with her valid credentials
  Then she sees a Free article and a Paid article on the home page
``` 

> With a declarative style, each step communicates an idea, but the exact values aren't specified. The details of "how" the user interacts with the system, such as which specific articles are free or paid, and the subscription level of different test users, are specified in the automation code that interacts with the system. The subscription packages can change in the future. The business can change what content is available to subscribers on free and paid plans, without having to change this test and other tests that use the same step definitions. If another subscription level is added later, it's easy to add a test for that. By avoiding terms like “click a button” that suggest implementation, the test is more resilient to implementation details of the UI. The intent of the test remains the same, even if the implementation changes later.

Reviewer: ```suggestion
With a declarative style, each step communicates an idea, but the exact values aren't specified. The details of *how* the user interacts with the system, such as which specific articles are free or paid, and the subscription level of different test users, are specified in the step definitions (the automation code that interacts with the system). The subscription packages could change in the future. The business could change what content is available to subscribers on free and paid plans, without having to change this scenario and other scenarios that use the same step definitions. If another subscription level is added later, it's easy to add a scenario for that. By avoiding terms like “click a button” that suggest implementation, the scenario is more resilient to implementation details of the UI. The intent of the scenario remains the same, even if the implementation changes later. In addition, having too many implementation details in a scenario, makes it harder to understand the intended behaviour it illustrates.
```

If the answer is “Yes”, then you should rework it avoiding implementation specific details.
As a side benefit, in consequence your scenarios will be a lot shorter and much easier to follow and understand.

# Consider a more declarative style

> One way to make tests easier to maintain and less brittle is to use a declarative style. Declarative style tests state ideas, rather than details of the UI. They are more informative, they can help tests read better as "living documentation". They are less brittle, and let the test have a longer lifespan. A declarative style helps you focus on the value that the customer is getting, rather than the keystrokes they will use,

Reviewer: "Declarative style tests state ideas, rather than details of the UI." -> To be more in line with "cucumber" / "bdd" terminology, I would say ""Declarative style describes the behaviour of the application, rather than the implementation details."

Reviewer: Sentence should end with a period `.` instead of a comma `,`

Author: Done.

> Imperative tests are communicative, but because they are so closely tied to the mechanics of the current UI, they are often more work to maintain. It's easiest to understand this by looking at examples.

Reviewer: "more work to maintain." -> Maybe add a sentence here to say "If the implementation changes, the tests need to be updated too" or something similar.

Author: Done.

> Here's a more imperative style;

Reviewer: Please use a `:` at the end of this sentence instead of `;` for consistency.

Author: Done.

```
Feature: Subscribers see different sets of stock images based on their subscription level 

Scenario: Free subscribers see only the free articles

> 

Reviewer: Please remove this blank line.

Author: Done.

```
  Given Free Frieda has a free subscription
  When Free Frieda logs in with her valid credentials
  Then she sees a Free article on the home page

  Given Paid Patty has a basic-level paid subscription
  When Paid Patty logs in with her valid credentials
  Then she sees a Free article and a Paid article on the home page
``` 
With a declarative style, each step communicates an idea, but the exact values aren't specified. The details, such as what specific articles are free or paid, and the subscription level of different test users, are specified "behind the scenes" in the step definition code. The subscription packages can change in the future. The business can change what content is available to subscribers on free and paid plans, without having to change this test and other tests that use the same step definitions. If another subscription level is added later, it's easy to add a test for that. By avoiding terms like “click a button” that suggest implementation, the test is more resilient to implementation details of the UI. The intent of the test remains the same, even if the implementation changes later.

> Imperative tests do have a place, but use them wisely. There’s no right or wrong, experiment to see what fits your context.  

Reviewer: Not sure about this paragraph. The rest of the content is all about being more declarative, so this feels out of place to me. Consider removing this section?

Author: I did remove it, but I added a bit on line 43 trying to convey that it is not *wrong* to use imperative style, it can be the best choice for some contexts. Is that ok?

Reviewer: Yes, it's much clearer now, thank you!
